### PR TITLE
Egat

### PR DESCRIPTION
**Pull request by manonreau**
1) Create edges feature by concatenating node feature
$e_{ij} = LeakyReLu (a_{ij} * [W * x_i || W * x_j])$

2) Apply softmax function, in order to learn to ignore some edges
$\alpha_{ij}  = softmax(e_{ij})$

3) Sum over the nodes (no averaging here !)
$z_i = \sum_j (\alpha_{ij} * Wx_j)$

Herein, we add the edge feature to the step 1)
$e_{ij} = LeakyReLu (a_{ij} * [W * x_i || W * x_j || We * edge_{attr} ])$

"""

### CODE REVIEW

File: graphprot/NeuralNet.py [hunk 1]
```diff
@@ -372,6 +374,7 @@ def test(self, database_test, threshold=4, hdf5='test_data.hdf5'):
 
         self.test_out = _out
         self.test_y = _y
+        print(_out, _y)
```
Comment (NicoRenaud) [inline graphprot/NeuralNet.py:377]: maybe remove the print :) With the code being serious now we should actually think of a logging system

File: graphprot/EGAT.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        ed = self.fc_edge_attr(edge_attr)
+        # create edge feature by concatenating node feature
+        a = torch.cat([xrow, xcol, ed], dim=1)
+        a = self.fc_attention(a)
+        e = F.leaky_relu(a)
+        
+        alpha = F.softmax(e, dim=1)
+        h = alpha * xcol 
+        
+        out = torch.zeros(
```
Comment (NicoRenaud) [inline graphprot/EGAT.py:64]: If you have a tensor of size `[num_node x out_channels]`  you can use 
```python 
out = torch.zeros_like(alpha)
```
That will directly create out of the same size than alpha but also directly on the same device. So that saves a host->device copy if we are running on GPU (I think)

File: graphprot/EGAT.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+
+        self.fc = nn.Linear(self.in_channels, self.out_channels, bias=bias)
+        self.fc_edge_attr = nn.Linear(1, 3, bias=bias)
+        self.fc_attention = nn.Linear(2 * self.out_channels + 3, 1, bias=bias)
+        self.reset_parameters()
+        
+    def reset_parameters(self):
```
Comment (NicoRenaud) [inline graphprot/EGAT.py:39]: `initialize_parameters` might be a better name ?

File: graphprot/EGAT.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+                 out_channels,
+                 bias=False):
+
+        super(EGAT, self).__init__()
+
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+
+        self.fc = nn.Linear(self.in_channels, self.out_channels, bias=bias)
+        self.fc_edge_attr = nn.Linear(1, 3, bias=bias)
```
Comment (NicoRenaud) [inline graphprot/EGAT.py:35]: so we can only have 1 edge attr ?

File: graphprot/EGAT.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        xrow = self.fc(x[row])
+        
+        ed = self.fc_edge_attr(edge_attr)
+        # create edge feature by concatenating node feature
+        a = torch.cat([xrow, xcol, ed], dim=1)
+        a = self.fc_attention(a)
+        e = F.leaky_relu(a)
+        
+        alpha = F.softmax(e, dim=1)
+        h = alpha * xcol
```
Comment (NicoRenaud) [inline graphprot/EGAT.py:62]: we could use only 2 variable instead of `a, e, alpha, h`. If we start using many graphs we might use more memory that we intend.

**Review by NicoRenaud — COMMENTED** [review]
It looks good ! I made a few comments. I didn't check really the implementation as I don't know exactly which paper you are referring to. If you send me the paper I can also check the operations
